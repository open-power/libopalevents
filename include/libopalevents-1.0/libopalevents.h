/*
 *  libopalevents - OPAL Event parsing and printing library
 *  Copyright (C) 2015 IBM Corporation

 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.

 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.

 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *  USA
 */

#ifndef _H_OPAL_EVENTS
#define _H_OPAL_EVENTS

#include <inttypes.h>

enum opal_event_log_flags {
	OPAL_EVENT_ALL = 0,
	OPAL_EVENT_ACTION = 1, /* Mutally exclusive */
	OPAL_EVENT_HEALTH = 2, /* Mutally exclusive */
	OPAL_EVENT_REPORT_ALL = 4,
	OPAL_EVENT_REPORT_HMC = 8,
	OPAL_EVENT_CALL_HOME = 16
};

typedef struct opal_event_log opal_event;

typedef enum opal_event_log_flags opal_event_flags;

opal_event *opal_event_create(void);

int opal_event_parse(opal_event *event, char *buf, int buflen);

int opal_event_print(opal_event *event);

int opal_event_summarise_hdr(void);

int opal_event_summarise(opal_event *event, opal_event_flags flags);

uint32_t opal_event_get_logid(opal_event *event);

int opal_event_destroy(opal_event *event);

#endif /* _H_OPAL_EVENTS */
