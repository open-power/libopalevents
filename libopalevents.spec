Name:           libopalevents
Version:        1
Release:        0%{?dist}
Summary:        PEL structure reader and parser

License:        LGPLv2
URL:            https://github.com/open-power/libopalevents
Source0:        https://github.com/open-power/libopalevents/archive/master.tar.gz
BuildRequires: autoconf automake libtool 
#Requires:       

%description 
PEL structure reader and parser
library to aid in reading, parsing and printing platform event log structures
 from files.



%package        devel
Summary:        Development files for %{name}
Requires:       %{name}%{?_isa} = %{version}-%{release}

%description    devel
The %{name}-devel package contains libraries and header files for
developing applications that use %{name}.


%prep

%setup -qn %{name}-master
autoreconf -if

%build
%configure --disable-static --prefix=/usr
make %{?_smp_mflags}


%install
rm -rf $RPM_BUILD_ROOT
%make_install
find $RPM_BUILD_ROOT -name '*.la' -exec rm -f {} ';'


%post -p /sbin/ldconfig

%postun -p /sbin/ldconfig


%files
%doc
%{_libdir}/*.so.*

%files devel
%doc
%{_includedir}/*
%{_libdir}/*.so
%{_libdir}/pkgconfig/libopalevents.pc


%changelog
* Wed Apr  8 2015 Cyril Bur <cyril.bur@au1.ibm.com> - 1
- Inital verion of the package
