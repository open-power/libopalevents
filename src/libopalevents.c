/*
 *  libopalevents - OPAL Event parsing and printing library
 *  Copyright (C) 2015 IBM Corporation

 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.

 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.

 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *  USA
 */

/*
 *  libopalevents - OPAL Event parsing and printing library
 *  Copyright (C) 2015 IBM Corporation

 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.

 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.

 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <inttypes.h>

#include "libopalevents-1.0/libopalevents.h"
#include "opal-event-log.h"
#include "parse-opal-event.h"
#include "print-opal-event.h"

opal_event *opal_event_create(void)
{
	return create_opal_event_log();
}

int opal_event_parse(opal_event *event, char *buf, int buflen)
{
	return parse_opal_event_log(buf, buflen, event);
}

int opal_event_print(opal_event *event)
{
	return print_opal_event_log(event);
}

int opal_event_summarise_hdr(void)
{
	return header_opal_event_log();
}

int opal_event_summarise(opal_event *event, opal_event_flags flags)
{
	return summarise_opal_event_log(event, flags);
}

uint32_t opal_event_get_logid(opal_event *event)
{
	return opal_event_log_id(event);
}

int opal_event_destroy(opal_event *event)
{
	return free_opal_event_log(event);
}
