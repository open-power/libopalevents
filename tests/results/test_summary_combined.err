parse_priv_hdr_scn: section header has an invalid section count 0, should be greater than 0, setting section count to 1 to attempt recovery
Warning parse_opal_event_log: detected missing section in opal event log
Warning parse_opal_event_log: detected missing section in opal event log
Warning parse_opal_event_log: detected missing section in opal event log
Warning parse_opal_event_log: detected opal event log sections out of order
Warning parse_opal_event_log: detected missing section in opal event log
ERROR parse_section_header: section header is corrupt. Length < 8 bytes and must be at least 8 bytes to include the length of itself. Id 0x00 Length 0 Version 0 Subtype 0 Component ID: 0
ERROR parse_opal_event_log: Could not parse section header, giving up
parse_priv_hdr_scn: section header has an invalid section count 0, should be greater than 0, setting section count to 1 to attempt recovery
Warning parse_opal_event_log: detected missing section in opal event log
parse_priv_hdr_scn: section header has an invalid section count 0, should be greater than 0, setting section count to 1 to attempt recovery
Warning parse_opal_event_log: detected missing section in opal event log
