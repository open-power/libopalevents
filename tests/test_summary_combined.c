#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <fcntl.h>
#include <unistd.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "libopalevents-1.0/libopalevents.h"
#include "test.h"

void test_summary_combined(void *cookie)
{
	int n, i = 0;
	struct dirent **files;

	opal_event *event;
	char *buf = NULL;
	off_t size;
	int ret;

	opal_event_summarise_hdr();
	n = scandir (TEST_INPUT_DIR, &files, NULL, alphasort);
	while (i < n) {
		int fsize = 512;
		char *file = malloc(fsize);
		if (!file)
			test_fail("Couldn't malloc filename\n");

		if (files[i]->d_type != DT_REG) {
			free(files[i]);
			i++;
			continue;
		}

		int oldfsize = fsize;
		if (fsize = snprintf(file, fsize, "%s%s", TEST_INPUT_DIR, files[i]->d_name) >= oldfsize) {
			file = malloc(fsize);
			assert(file);
			snprintf(file, fsize, "%s%s", TEST_INPUT_DIR, files[i]->d_name);
		} else {
			fsize = oldfsize;
		}
		size = test_open_file(file, &buf);

		event = opal_event_create();

		opal_event_parse(event, buf, size);
		ret = opal_event_summarise(event, OPAL_EVENT_ALL);
		if (ret)
			printf("|Couldn't opal_event_summarise for file %s ret=%d              |\n", files[i]->d_name, ret);

		opal_event_destroy(event);

		/*
		 * Important to set it to NULL so that when the test suite frees it
		 * nothing bad happens
		 */
		free(buf);
		buf = NULL;

		free(files[i]);
		free(file);
		i++;
	}

	if (n < 0) {
		/* could not open drectory */
		test_fail("Could not open test file directory '%s'", TEST_INPUT_DIR);
	} else {
		free(files);
	}

	test_assert_diff();

	test_passed();
}


int main(int argc, char *argv[])
{
	TEST the_t = {
		.name = "test_summary_combined",
		.t = &test_summary_combined,
		.cookie = (void *) 0, /* Unused */
		.r = { "test_summary_combined.out", "test_summary_combined.err" },
		.i = { "" } /* Unused */
	};

	return test_main(the_t);
}
