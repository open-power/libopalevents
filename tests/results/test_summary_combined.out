|------------------------------------------------------------------------------|
|ID       Date       Time     SRC        Creator           Event Severity      |
|------------------------------------------------------------------------------|
|00000003 2014-03-13 13:01:56 ?UNKNOWN ? Unknown           ?UNKNOWN            |
|50000004 2000-12-31 10:14:44 ?UNKNOWN ? OPAL              ?UNKNOWN            |
|50000006 2014-03-14 14:37:00 ?UNKNOWN ? OPAL              ?UNKNOWN            |
|5034A000 2014-03-13 08:15:55 11007201 + Service Processor Predictive Error    |
|5055ED2E 2014-02-18 06:43:54 B182950C   Service Processor Informational Event |
|Couldn't opal_event_summarise for file test_parse_000.raw ret=-1              |
|00000002 0000-00-00 00:00:00 ?UNKNOWN ? Unknown           ?UNKNOWN            |
|00000002 0000-00-00 00:00:00 ?UNKNOWN ? Unknown           ?UNKNOWN            |
